src/tracker/mime-type-tracker-query.vala

```diff
@@ -34,9 +38,31 @@ public class Games.MimeTypeTrackerQuery : Object, TrackerQuery {
 		return false;
 	}
 
-	public Game game_for_cursor (Tracker.Sparql.Cursor cursor) throws Error {
-		var uri = cursor.get_string (0);
+	public void process_cursor (Tracker.Sparql.Cursor cursor) {
+		 var uri = cursor.get_string (0);
+		 uris_for_source += uri;
+	}
+
+	public async void foreach_game (GameCallback game_callback) {
+		uint handled_games = 0;
+		for (size_t i = 0; i < uris_for_source.length; i++) {
+			var game = game_for_cursor (uris_for_source[i]);
+			game_callback (game);
+			handled_games++;
+
+			// Free the execution only once every HANDLED_GAMES_PER_CYCLE
+			// games to speed up the execution by avoiding too many context
+			// switching.
+			if (handled_games >= HANDLED_GAMES_PER_CYCLE) {
+				handled_games = 0;
+
+				Idle.add (this.foreach_game.callback);
+				yield;
+			}
+		}
+	}
 
+	public Game game_for_cursor (string uri) throws Error {
```

Comment:
It collides with the callback in this very class. I can rename it to game_for_uri_cb